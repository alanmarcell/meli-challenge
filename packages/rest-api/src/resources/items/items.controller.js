import axios from 'axios';

const errorHandler = (res, error) => res.status(500).json({ message: error.message });


const instance = axios.create({
  baseURL: 'https://api.mercadolibre.com',
  // headers: { 'Content-Type': 'application/json' },
});

const getItemsService = query => {

  return instance.get(`/sites/MLA/search?q=${query}`)
}

const getItems = async (req, res, _next) => {
  try {
    console.log(req.query.q)
    const response = await getItemsService(req.query.q)
    // console.log(response)
    res.json(response.data);
  } catch (error) {
    errorHandler(res, error);
  }
}

export { getItems }