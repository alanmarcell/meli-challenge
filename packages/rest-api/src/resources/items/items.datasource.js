import axios from 'axios';

const instance = axios.create({
  baseURL: 'https://api.mercadolibre.com',
});

const getItemsService = (query) => instance.get(`/sites/MLA/search?q=${query}`);
const getItemService = (id) => instance.get(`/items/${id}`);
const getItemDescriptionService = (id) => instance.get(`/items/${id}/description`);

export {
  instance,
  getItemsService,
  getItemService,
  getItemDescriptionService,
};
