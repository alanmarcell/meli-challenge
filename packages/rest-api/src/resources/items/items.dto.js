import * as R from 'ramda';

const getCategories = (itemsRaw) => R.map((value) => value.name,
  R.prop('values',
    R.find((availableFilter) => availableFilter.id === 'category',
      R.prop('available_filters', itemsRaw))));

const itemsDto = (itemsRaw) => {
  const categories = getCategories(itemsRaw);

  const items = R.map((result) => {
    const amount = R.split('.', result.price.toString());

    return {
      id: result.id,
      title: result.title,
      price: {
        currency: result.currency_id,
        amount: Number(amount[0]),
        decimals: amount[1] ? Number(amount[1]) : 0,
      },
      picture: result.thumbnail,
      condition: result.condition,
      free_shipping: result.shipping.free_shipping,
    };
  }, R.prop('results', itemsRaw));

  return {
    author: {
      name: 'Alan',
      lastname: 'Marcell',
    },
    categories,
    items,
  };
};

export { itemsDto };
