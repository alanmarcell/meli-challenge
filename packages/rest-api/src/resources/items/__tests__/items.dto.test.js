import { itemsDto } from '../items.dto';
import rawItemsMock from './items.mock';

describe('Items DTO', () => {
  it('should have correct format', () => {
    expect(itemsDto(rawItemsMock)).toMatchSnapshot();
  });
});
