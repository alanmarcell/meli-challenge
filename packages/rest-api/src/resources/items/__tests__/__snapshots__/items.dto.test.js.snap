// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Items DTO should have correct format 1`] = `
Object {
  "author": Object {
    "lastname": "Marcell",
    "name": "Alan",
  },
  "categories": Array [
    "Consolas",
    "Libros",
    "Videojuegos",
    "Juegos y Juguetes",
    "Ropa y Accesorios",
    "Música, Películas y Series",
    "Hogar, Muebles y Jardín",
    "Accesorios para Consolas",
    "Computación",
    "Electrónica, Audio y Video",
    "Arte, Librería y Mercería",
    "Antigüedades y Colecciones",
    "Revistas",
    "Souvenirs, Cotillón y Fiestas",
    "Comics e Historietas",
    "Cámaras y Accesorios",
    "Herramientas y Construcción",
    "Joyas y Relojes",
    "Manga",
    "Otros",
    "Otras categorías",
  ],
  "items": Array [
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA653546463",
      "picture": "http://mla-s2-p.mlstatic.com/717025-MLA25351455547_022017-I.jpg",
      "price": Object {
        "amount": 1995,
        "currency": "ARS",
        "decimals": 85,
      },
      "title": "Consola De Video Juego Family Alien 1",
    },
    Object {
      "condition": "used",
      "free_shipping": false,
      "id": "MLA689525341",
      "picture": "http://mla-s2-p.mlstatic.com/906507-MLA26247945564_102017-I.jpg",
      "price": Object {
        "amount": 15000,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Videos Juegos Arcade",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA833264006",
      "picture": "http://mla-s1-p.mlstatic.com/641529-MLA40341969944_012020-I.jpg",
      "price": Object {
        "amount": 15899,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Nintendo Famicom Original + Opcional Reforma Video Gratis ",
    },
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA822812938",
      "picture": "http://mla-s1-p.mlstatic.com/623231-MLA32724407391_102019-I.jpg",
      "price": Object {
        "amount": 1299,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Arcade Mame Fichines Gabinete Mini Video Juegos Retro Maquin",
    },
    Object {
      "condition": "used",
      "free_shipping": false,
      "id": "MLA749659195",
      "picture": "http://mla-s2-p.mlstatic.com/966715-MLA28224423490_092018-I.jpg",
      "price": Object {
        "amount": 4160,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Consola Antigua Video Computer 2600 Edu Games",
    },
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA763348812",
      "picture": "http://mla-s1-p.mlstatic.com/937923-MLA31572885682_072019-I.jpg",
      "price": Object {
        "amount": 399,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Tetris Portátil Brick Game Con 9999 Video Juegos Clásicos",
    },
    Object {
      "condition": "used",
      "free_shipping": false,
      "id": "MLA828816784",
      "picture": "http://mla-s1-p.mlstatic.com/10396-MLA20028056363_012014-I.jpg",
      "price": Object {
        "amount": 49000,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Videos Games, Juegos,time Crisis 2, La Maquina Mas Jugadora",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA834691132",
      "picture": "http://mla-s1-p.mlstatic.com/757147-MLA40451818275_012020-I.jpg",
      "price": Object {
        "amount": 13432,
        "currency": "ARS",
        "decimals": 68,
      },
      "title": "Elgato Capturadora Game Streaming Hd60 S Usb 3.0",
    },
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA644964944",
      "picture": "http://mla-s1-p.mlstatic.com/749865-MLA31040358657_062019-I.jpg",
      "price": Object {
        "amount": 99,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Cable Salida A Tv Rf Family Game Sega Video Casetera Rosario",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA834694299",
      "picture": "http://mla-s2-p.mlstatic.com/768943-MLA41022111649_032020-I.jpg",
      "price": Object {
        "amount": 14798,
        "currency": "ARS",
        "decimals": 57,
      },
      "title": "Elgato Capturadora Game Streaming Hd 60 Pro",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA834446952",
      "picture": "http://mla-s1-p.mlstatic.com/936654-MLA40435392527_012020-I.jpg",
      "price": Object {
        "amount": 10659,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Bus Simulator - Playstation 4 (fisico) ",
    },
    Object {
      "condition": "used",
      "free_shipping": false,
      "id": "MLA821629029",
      "picture": "http://mla-s2-p.mlstatic.com/914072-MLA32645907246_102019-I.jpg",
      "price": Object {
        "amount": 4588,
        "currency": "ARS",
        "decimals": 5,
      },
      "title": "Antigua Consola Tipo Telematch Concept 2000 4 Juegos Aps",
    },
    Object {
      "condition": "used",
      "free_shipping": false,
      "id": "MLA844952178",
      "picture": "http://mla-s1-p.mlstatic.com/952274-MLA41198562234_032020-I.jpg",
      "price": Object {
        "amount": 550,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Video Juego Tetris A Pilas Retro Años 90 Excelente Estado",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA787502508",
      "picture": "http://mla-s1-p.mlstatic.com/917350-MLA31634902309_072019-I.jpg",
      "price": Object {
        "amount": 14699,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Capturadora Elgato Game Capture Hd60 Pro ",
    },
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA813579669",
      "picture": "http://mla-s2-p.mlstatic.com/894036-MLA32113911507_092019-I.jpg",
      "price": Object {
        "amount": 2207,
        "currency": "ARS",
        "decimals": 96,
      },
      "title": "Consola Video Juego Retro Gameboy Arcade Family Vintage Mame",
    },
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA726626875",
      "picture": "http://mla-s2-p.mlstatic.com/903204-MLA32094765905_092019-I.jpg",
      "price": Object {
        "amount": 69,
        "currency": "ARS",
        "decimals": 9,
      },
      "title": "Cable De Audio Estereo Plug 3.5mm A 3.5mm Mallado 1mt",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA739022071",
      "picture": "http://mla-s2-p.mlstatic.com/836714-MLA27827110980_072018-I.jpg",
      "price": Object {
        "amount": 3399,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Lego Batman The Videogame Ps3 Juego Original Fisico Sellado",
    },
    Object {
      "condition": "used",
      "free_shipping": false,
      "id": "MLA672749256",
      "picture": "http://mla-s1-p.mlstatic.com/690978-MLA25586035861_052017-I.jpg",
      "price": Object {
        "amount": 9500,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Consola Retroplay Raspberry Pi3 Dreamcast -ver Video- Sjl",
    },
    Object {
      "condition": "used",
      "free_shipping": false,
      "id": "MLA768191432",
      "picture": "http://mla-s1-p.mlstatic.com/873318-MLA29241947196_012019-I.jpg",
      "price": Object {
        "amount": 14000,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Arcades Video Juegos Clasicos",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA786212862",
      "picture": "http://mla-s1-p.mlstatic.com/781697-MLA31355988977_072019-I.jpg",
      "price": Object {
        "amount": 15399,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Capturadora Game Streaming Elgato Hd 60 Pro",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA763059758",
      "picture": "http://mla-s1-p.mlstatic.com/688975-MLA29024227573_122018-I.jpg",
      "price": Object {
        "amount": 5899,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Monster Energy Supercross - The Official Videogame 2 -...",
    },
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA726628858",
      "picture": "http://mla-s1-p.mlstatic.com/643003-MLA41243677909_032020-I.jpg",
      "price": Object {
        "amount": 79,
        "currency": "ARS",
        "decimals": 99,
      },
      "title": "Cable Auxiliar Stereo 1m Audio Mini Plug 3.5mm Mallado",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA763064898",
      "picture": "http://mla-s1-p.mlstatic.com/662173-MLA31604178578_072019-I.jpg",
      "price": Object {
        "amount": 19069,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Lego Dc Super-villains Deluxe Edition - Playstation 4",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA763036400",
      "picture": "http://mla-s1-p.mlstatic.com/669877-MLA29023337561_122018-I.jpg",
      "price": Object {
        "amount": 6769,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Rocket League Ultimate Edition - Playstation 4",
    },
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA813260287",
      "picture": "http://mla-s2-p.mlstatic.com/721408-MLA41243670829_032020-I.jpg",
      "price": Object {
        "amount": 79,
        "currency": "ARS",
        "decimals": 9,
      },
      "title": "Cable Mini Plug Miniplug 3,5mm Mallado Conect Metal 1mt Aux4",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA739044462",
      "picture": "http://mla-s1-p.mlstatic.com/996631-MLA31021139252_062019-I.jpg",
      "price": Object {
        "amount": 3199,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Lego The Movie Videogame Ps3 Cd Original Fisico Sellado",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA763036335",
      "picture": "http://mla-s2-p.mlstatic.com/707103-MLA29023334729_122018-I.jpg",
      "price": Object {
        "amount": 4979,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Monster Energy Supercross - The Official Videogame 2 - Xbox",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA799999945",
      "picture": "http://mla-s2-p.mlstatic.com/723226-MLA31442895784_072019-I.jpg",
      "price": Object {
        "amount": 4499,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Dvd Portatil 7,8  Usb Game Mp3",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA765698057",
      "picture": "http://mla-s2-p.mlstatic.com/987126-MLA29134646058_012019-I.jpg",
      "price": Object {
        "amount": 2850,
        "currency": "ARS",
        "decimals": 86,
      },
      "title": "Lego Movie Ps3 Fisico",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA756815918",
      "picture": "http://mla-s1-p.mlstatic.com/935300-MLA28586478094_112018-I.jpg",
      "price": Object {
        "amount": 3034,
        "currency": "ARS",
        "decimals": 79,
      },
      "title": "Lego Ninjago Movie Videogame - Playstation 4",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA802050306",
      "picture": "http://mla-s1-p.mlstatic.com/664743-MLA31511751411_072019-I.jpg",
      "price": Object {
        "amount": 5359,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "N3ii: Ninety-nine Nights - Xbox 360",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA825283893",
      "picture": "http://mla-s2-p.mlstatic.com/872090-MLA32878955729_112019-I.jpg",
      "price": Object {
        "amount": 48449,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Blu-ray : Game Of Thrones: Complete Series Gift Set (33...",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA693469984",
      "picture": "http://mla-s1-p.mlstatic.com/682969-MLA26432246747_112017-I.jpg",
      "price": Object {
        "amount": 2574,
        "currency": "ARS",
        "decimals": 97,
      },
      "title": "Lego Batman The Videogame With Batman Movie Combo Pack Ps3",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA815437483",
      "picture": "http://mla-s2-p.mlstatic.com/773437-MLA32215526351_092019-I.jpg",
      "price": Object {
        "amount": 2565,
        "currency": "ARS",
        "decimals": 97,
      },
      "title": "Lego The Movie Videogame Nuevo Sellado Fisico Ps4",
    },
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA681282439",
      "picture": "http://mla-s2-p.mlstatic.com/648616-MLA32404132921_102019-I.jpg",
      "price": Object {
        "amount": 999,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Cable Hdmi Premium 2mts Ps3 Ps4 Xbox Pc 1080p 4k Game",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA773458260",
      "picture": "http://mla-s1-p.mlstatic.com/703426-MLA29533942195_032019-I.jpg",
      "price": Object {
        "amount": 2747,
        "currency": "ARS",
        "decimals": 58,
      },
      "title": "Lego The Movie Videogame Nuevo Nintendo 3ds",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA665426522",
      "picture": "http://mla-s1-p.mlstatic.com/951515-MLA25572791277_052017-I.jpg",
      "price": Object {
        "amount": 2614,
        "currency": "ARS",
        "decimals": 91,
      },
      "title": "The Lego Movie Videogame - Nintendo 3ds",
    },
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA803900282",
      "picture": "http://mla-s1-p.mlstatic.com/978651-MLA31581241607_072019-I.jpg",
      "price": Object {
        "amount": 1199,
        "currency": "ARS",
        "decimals": 99,
      },
      "title": "Figura Pikachu Pokemon Pop En Caja Jugueteria Medrano 21",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA780780044",
      "picture": "http://mla-s1-p.mlstatic.com/821397-MLA31620397160_072019-I.jpg",
      "price": Object {
        "amount": 2799,
        "currency": "ARS",
        "decimals": 99,
      },
      "title": "Lego Movie 2 Ps4 Físico - Nuevo",
    },
    Object {
      "condition": "used",
      "free_shipping": false,
      "id": "MLA846561475",
      "picture": "http://mla-s2-p.mlstatic.com/667179-MLA31656432109_082019-I.jpg",
      "price": Object {
        "amount": 690,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Uncharted 3 Ps3 Fisico Usado",
    },
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA614276486",
      "picture": "http://mla-s1-p.mlstatic.com/970587-MLA31048744171_062019-I.jpg",
      "price": Object {
        "amount": 449,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Cortina De Baño Pacman +protector + Ganchos. Super Oferta!",
    },
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA845237401",
      "picture": "http://mla-s2-p.mlstatic.com/691123-MLA41226636521_032020-I.jpg",
      "price": Object {
        "amount": 199,
        "currency": "ARS",
        "decimals": 99,
      },
      "title": "Lego Movie | Ps3 | Playstation 3 - Tenelo Hoy!",
    },
    Object {
      "condition": "used",
      "free_shipping": false,
      "id": "MLA847109408",
      "picture": "http://mla-s1-p.mlstatic.com/669245-MLA28625489015_112018-I.jpg",
      "price": Object {
        "amount": 799,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Gta 4 Liberty City Ps3 Fisico Usado Grand Theft Auto Iv ",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA846190796",
      "picture": "http://mla-s1-p.mlstatic.com/720470-MLA41284867639_032020-I.jpg",
      "price": Object {
        "amount": 19344,
        "currency": "ARS",
        "decimals": 21,
      },
      "title": "Elgato Capturadora Game Streaming Capture 4k60 Pro",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA773155610",
      "picture": "http://mla-s1-p.mlstatic.com/881837-MLA29517158053_022019-I.jpg",
      "price": Object {
        "amount": 3499,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Lego Movie 2 Videogames Ps4 Fisico Nuevo Xstation Estreno!",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA817473963",
      "picture": "http://mla-s2-p.mlstatic.com/839984-MLA32376872486_092019-I.jpg",
      "price": Object {
        "amount": 3499,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "The Lego Movie 2 Nintendo Switch Fisico Ps4 Games San Justo",
    },
    Object {
      "condition": "new",
      "free_shipping": false,
      "id": "MLA759933821",
      "picture": "http://mla-s2-p.mlstatic.com/974376-MLA31115978966_062019-I.jpg",
      "price": Object {
        "amount": 1995,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Funko Pop Games Spider-man Unmasked",
    },
    Object {
      "condition": "used",
      "free_shipping": false,
      "id": "MLA731888676",
      "picture": "http://mla-s1-p.mlstatic.com/690359-MLA27573388663_062018-I.jpg",
      "price": Object {
        "amount": 1500,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Wasteland 2, Físico Usado Ps4 Crash Video Games",
    },
    Object {
      "condition": "used",
      "free_shipping": false,
      "id": "MLA838575843",
      "picture": "http://mla-s2-p.mlstatic.com/672406-MLA40730645152_022020-I.jpg",
      "price": Object {
        "amount": 2750,
        "currency": "ARS",
        "decimals": 0,
      },
      "title": "Canje 1x1 Ghostbusters/cazafantasmas Ps3 Venta Envío S/cargo",
    },
    Object {
      "condition": "new",
      "free_shipping": true,
      "id": "MLA749425649",
      "picture": "http://mla-s2-p.mlstatic.com/685726-MLA31045573115_062019-I.jpg",
      "price": Object {
        "amount": 3857,
        "currency": "ARS",
        "decimals": 31,
      },
      "title": "Neca Terminator 2 T800 Video Game Appearance Original Oferta",
    },
  ],
}
`;
